clippy_lints/src/missing_const_for_fn.rs

```diff
@@ -185,3 +190,17 @@ fn could_be_const_with_abi(cx: &LateContext<'_>, msrv: &Msrv, abi: Abi) -> bool
         _ => cx.tcx.features().const_extern_fn,
     }
 }
+
+/// Return `true` when the given `def_id` is a function that has `impl Trait` ty as one of
+/// its parameter types.
+fn fn_inputs_has_impl_trait_ty(cx: &LateContext<'_>, def_id: LocalDefId) -> bool {
+    let inputs = cx.tcx.fn_sig(def_id).instantiate_identity().inputs().skip_binder();
+    inputs.iter().any(|input| {
+        // NB: Other alias ty kind might missing default.
+        // For example, an associate type alias declared as `type T: Fn();`
+        // would cause ICE when `type_of` is called with it.
```

Comment:
I'm not sure what this comment means, I've tested if changing test's Foo to that piece of code caused an ICE (that would be quite the blocker) and it doesn't cause one.

Reply:
Oh, what I meant was:

When matching the `input.kind()` on the next line, if the alias ty is something else, i.e. `AliasTyKind::Projection` would cause ICE in that scenario. You can verify it by changing the `ty::AliasTyKind::Weak` to `_`  as well.

(It seems that it only happens with `AliasTyKind::Projection` atm, but I wasn't too sure, so I put the comment there trying to give a head's up for others)

I'll make some adjustments on this, along with the test case~

Reply:
wait, as I adding a test to trigger ICE, I accidently found a case that triggers issue #13013 , might as well try to fix that as well, hope you wouldn't mind~

@rustbot author